### PR TITLE
Separate Main Loop

### PR DESCRIPTION
**Pull request by kphoenix137**
Based on: https://github.com/diasurgical/devilutionx-gamelist/pull/18
Fixes: https://github.com/diasurgical/devilutionx-gamelist/issues/17

### CODE REVIEW

**Comment by AJenbo** [comment]
There are a few small conflicts, could you please resolve them?

**Comment by kphoenix137** [comment]
> There are a few small conflicts, could you please resolve them?

Should be all set.

**Review by AJenbo — DISMISSED** [review]
Thanks this is really exciting!

**Comment by AJenbo** [comment]
Looks like this change causes it to get stuck here:
image

I'll try to debug it a bit alter

File: discord_bot.py [hunk 1]
```diff
@@ -203,12 +205,25 @@ def any_player_name_contains_a_banned_word(players):
 background_task_running = 0
 
 
+async def main_task():
+    while True:
+        if client.is_ready():  # This ensures the client is ready before starting the background task.
```
Comment (StephenCWills) [inline discord_bot.py:210]: Does the Discord SDK call `on_ready()` when the client is not ready?

If not, then it seems like you're implying that the readiness can come and go. Does the SDK not call `on_ready()` again when the client is made ready again?

Comment (AJenbo) [reply discord_bot.py:210]: That's also the understanding I have. Some times this happens, and this is usually where the old code dies:

image

Comment (StephenCWills) [reply discord_bot.py:210]: Looks like there's different events. `on_ready()` may be called again if a `RESUME` request fails, based on their reconnection logic.

https://discordpy.readthedocs.io/en/latest/api.html#gateway

Comment (StephenCWills) [reply discord_bot.py:210]: > ... this is usually where the old code dies

Does that screenshot suggest that the code dies when the `RESUME` request succeeds? And that the bot is actually still running, but not updating the Discord channel?

File: discord_bot.py [hunk 2]
```diff
@@ -267,14 +276,34 @@ async def background_task():
 
             activity = discord.Activity(name='Games online: '+str(current_online), type=discord.ActivityType.watching)
             await client.change_presence(activity=activity)
+    background_task_running = False
```
Comment (StephenCWills) [inline discord_bot.py:279]: You put this right after an infinite loop. Is it possible to reach this line of code?

Comment (kphoenix137) [reply discord_bot.py:279]: Good catch. No

File: discord_bot.py (hunk 1 above)
Comment (AJenbo) [reply discord_bot.py]: No that was an example of it succesfuld resuming.

Here is an example where it stopped working:
image

It's currently pretty unclear to me what will cause either.

File: discord_bot.py [hunk 3]
```diff
@@ -200,81 +202,105 @@ def any_player_name_contains_a_banned_word(players):
...
 
 @client.event
 async def on_ready():
     print(f'We have logged in as {client.user}')
     global global_channel
     global_channel = client.get_channel(DISCORD_CHANNEL_ID)
-    await background_task()
+
+    loop = asyncio.get_event_loop()
+    loop.create_task(background_task())  # Run background_task() as a separate task when the client is ready.
```
Comment (StephenCWills) [inline discord_bot.py:302]: Is it necessary to call `create_task()` on the current event loop, or can we just call the higher-level `asyncio.create_task()` function? The asyncio documentation suggests that the event loop API calls are low-level functions that typically shouldn't need to be used by applications.

https://docs.python.org/3/library/asyncio-eventloop.html

Comment (AJenbo) [reply discord_bot.py:302]: I have no clue

File: discord_bot.py (hunk 1 above)
Comment (StephenCWills) [reply discord_bot.py]: There is an `on_error()` event that we can use to get some feedback from uncaught exceptions. However, there is a note in the documentation that suggests we should have already been receiving traceback logs to stderr. Could it be that we're stuck in an infinite loop or a deadlock?

https://discordpy.readthedocs.io/en/latest/api.html#discord.on_error

**Pull request closed without merging**